modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java

```diff
@@ -65,37 +62,24 @@
 			_frontendTokenDefinitionRegistry.getFrontendTokenDefinition(
 				layoutSet.getThemeId());
 
-		ParsedFrontendTokenDefinition parsedFrontendTokenDefinition =
-			frontendTokenDefinition.getParsedFrontendTokenDefinition(
-				themeDisplay.getLocale());
+		Collection<FrontendTokenMapping> frontendTokenMappings =
+			frontendTokenDefinition.getFrontendTokenMappings();
 
-		Collection<FrontendTokenCategory> frontendTokenCategories =
-			parsedFrontendTokenDefinition.getFrontendTokenCategories();
+		for (FrontendTokenMapping frontendTokenMapping :
+				frontendTokenMappings) {
 
-		for (FrontendTokenCategory frontendTokenCategory :
-				frontendTokenCategories) {
+			String type = frontendTokenMapping.getType();
 
-			for (FrontendTokenSet frontendTokenSet :
-					frontendTokenCategory.getFrontendTokenSets()) {
-
-				for (FrontendToken frontendToken :
-						frontendTokenSet.getFrontendTokens()) {
-
-					for (FrontendTokenMapping frontendTokenMapping :
-							frontendToken.getFrontendTokenMappings()) {
-
-						String type = frontendTokenMapping.getType();
+			if (!type.equals("cssVariable")) {
+				continue;
+			}
 
-						if (!type.equals("cssVariable")) {
-							continue;
-						}
+			FrontendToken frontendToken =
```

Comment:
This inversion of control feels a bit weird... wouldn't it feel more natural to loop through `Token` asking for the mappings?

One convenience method could be `FrontendTokenMapping cssVariableMapping = token.getMapping(FrontendTokenMapping.CSS_VARIABLE)`.

Reply:
This is assuming only one type of mapping, but it could be easily generalized to a Collection.

Reply:
I don't see any inversion of control: we have methods to traverse the structure up or down the way you like, and then methods to get all descendants directly (jumping several levels down) so I do it that way to avoid an extra `for`. However, if we add another method to filter by mapping type it can be done with only one `for` as you suggest, of course. 

But, given that we haven't defined mapping types, I'm not sure we want to add that method now. That leads us to another question: are the mapping types defined? If so, where (the JSON schema is missing them)? Seeing the code I thought it was left open for now (and maybe forever, I'm not sure), but maybe that assumption is wrong... :thinking:

Reply:
By "inversion" I meant taking a seemingly isolated "mapping" and then asking for its parent, rather than getting the parents and then querying downwards.

Even if we don't want to have the `filter`, I'd still prefer to get the `tokens` and then go through the mappings. Feels more natural and easier to reason with in the future. Also in case we need something else from the token, we already have the root that we care about in the first step.

Reply:
Thinking more about the filtering I think it's OK, but we need to decide if the mapping type is a `String` or an `enum`...

If it is a string then nobody defines its possible values (this creates a "problem" when using `cssVariable` because both Echo and us agree on using that...). 

If it is an enum, we (frontend infra) define its valid values, but then it cannot be extended (which can be a limitation or a feature).

Which one seems to be correct? :thinking: 

On the one hand, when themes are used, it looks like mapping types should be normalized (because adding a new type in, say, collaboration, won't work for anything as the base themes won't have that type of mappings). But on the other hand, if we allow using any arbitrary string, people can create their own themes with their own mapping types (for example: for mobile UIs, for any UI system they have, etc...).

I'm not sure what to do because I still haven't 100% clear what other use cases we have in mind...

Reply:
I'd say just `String` for now should be okay. Maybe just expose the `CSS_VARIABLE` string as a constant so we can all have a safe way to refer to it?

Reply:
Given that we are the ones "defining" the CSS variables thing (because it's in **OUR** themes) it makes sense that we define the `CSS_VARIABLE` constant. 

However, I don't think `frontend-token-definitions-api` is a good place for it, because it is far from themes.

But: I can't think of any better place, so I'd say let's put it inside `FrontendTokenMapping` for now.

Maybe the purest way to do it would be adding some OSGi module to hold the `DefaultThemeScopedCSSVariablesProvider` and the constant, that takes care of the frontend tokens stuff related to themes and css variable, but looks a bit overkill (at least for now; and also is quite unclear what will happen in the future, I think).

To sum up: I will put it in `FrontendTokenMapping` unless you can think of a better way to do it.

Reply:
Sounds good to me